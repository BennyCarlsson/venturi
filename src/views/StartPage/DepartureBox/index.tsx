import Heading, { HeadingType } from 'components/Heading';
import Paragraph from 'components/Paragraph';
import colors from 'tokens/colors';
import { FontWeight } from 'types';
import {
  BottomPart,
  BottomParagraph,
  BottomLeftPart,
  BottomRightPart,
  DepartsWrapper,
  NextDepartureWrapper,
  StyledTrailLogo,
  TopPart
} from './styles';

type NextDepartureBoxProps = {
  name: string | undefined;
  track: string | undefined;
};

const DepartureBox = (props: NextDepartureBoxProps) => {
  const { name, track } = props;
  return (
    <NextDepartureWrapper>
      <TopPart>
        <Heading
          headingType={HeadingType.H1}
          fontWeight={FontWeight.medium}
          fontSize={18}
          color={colors.white}
        >
          Departs from
        </Heading>
        <DepartsWrapper>
          {/* Todo: icon after type */}
          <StyledTrailLogo />
          <Heading
            headingType={HeadingType.H2}
            fontWeight={FontWeight.medium}
            fontSize={28}
            color={colors.white}
          >
            {name}
          </Heading>
        </DepartsWrapper>
      </TopPart>
      <BottomPart>
        <BottomLeftPart>
          <Paragraph fontWeight={FontWeight.medium} fontSize={14} color={colors.white}>
            Platform
          </Paragraph>
          <BottomParagraph fontWeight={FontWeight.medium} fontSize={24} color={colors.white}>
            {track}
          </BottomParagraph>
        </BottomLeftPart>
        <BottomRightPart>
          <Paragraph fontWeight={FontWeight.medium} fontSize={14} color={colors.white}>
            Destination name
          </Paragraph>
          <BottomParagraph fontWeight={FontWeight.medium} fontSize={24} color={colors.white}>
            Placeholder
          </BottomParagraph>
        </BottomRightPart>
      </BottomPart>
    </NextDepartureWrapper>
  );
};

export default DepartureBox;
