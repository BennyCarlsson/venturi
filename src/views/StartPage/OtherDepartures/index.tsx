import Paragraph from 'components/Paragraph'
import Colors from 'tokens/colors'
import { FontWeight } from 'types'
import { Trip } from 'redux/tripSlice'
import ChevronIcon from './ChevronIcon'
import {
  Box,
  DepartureWrapper,
  NameBox,
  TimeAndDirectionWrapper,
  DirectionParagraph,
  TitleParagraph,
  Top,
  Content,
} from './styles'

type Departures = {
  handleOnClick: () => void
  showContent: boolean
  trips: Trip[]
}

const OtherDepartures = ({ handleOnClick, showContent, trips }: Departures) => {
  const variant = {
    initial: { height: 0 },
    animate: {
      height: showContent ? 'fit-content' : 0,
      transition: { duration: 0.3 },
    },
  }

  const chevron = {
    animate: { rotate: showContent ? 180 : 0, transition: { duration: 0.2 } },
  }

  return (
    <Box>
      <Top
        onClick={(e) => {
          e.preventDefault()
          handleOnClick()
        }}
      >
        <span />
        <TitleParagraph
          center
          color={Colors.white}
          fontSize={18}
          fontWeight={FontWeight.medium}
        >
          Other departures
        </TitleParagraph>
        <ChevronIcon variants={chevron} animate={'animate'} />
      </Top>

      <Content variants={variant} animate={'animate'} initial={'initial'}>
        {trips.map((trip) => {
          //Todo leg can be an array of legs if trip contains multiple steps
          return (
            <DepartureWrapper key={trip.id}>
              <TimeAndDirectionWrapper>
                <Paragraph fontSize={18} color={Colors.white}>
                  {trip.departureTime}
                </Paragraph>
                <DirectionParagraph fontSize={18} color={Colors.white}>
                  {trip.direction}
                </DirectionParagraph>
              </TimeAndDirectionWrapper>
              <NameBox>
                <Paragraph fontSize={14}>{trip.number}</Paragraph>
              </NameBox>
            </DepartureWrapper>
          )
        })}
      </Content>
    </Box>
  )
}

export default OtherDepartures
