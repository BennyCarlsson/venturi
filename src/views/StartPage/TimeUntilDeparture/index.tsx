import Paragraph from "components/Paragraph";
import Colors from "tokens/colors";
import { FontWeight } from "types";

type TimeUntilDepartureProps = {
  date?: string;
  time?: string;
  rtime?: string;
};

const TimeUntilDeparture = ({ time, date }: TimeUntilDepartureProps) => {
  const getDays = (ms: number) => {
    const days = Math.floor(ms / 86400000);
    if (days <= 0) return "";
    if (days === 1) return `${days} day `;
    return `${days} days `;
  };

  const getHours = (ms: number) => {
    const hours = Math.floor((ms % 86400000) / 3600000);
    if (hours <= 0) return "";
    if (hours === 1) return `${hours} hour `;
    return `${hours} hours `;
  };

  const getMinutes = (ms: number) => {
    const minutes = Math.round(((ms % 86400000) % 3600000) / 60000);
    if (minutes <= 0) return "";
    if (minutes === 1) return `${minutes} minute `;
    return `${minutes} minutes `;
  };

  const getTimeUntilDeparture = () => {
    if (!date || !time) return;
    const today = new Date();
    var arr: any[] = `${date} ${time}`.split(/[- :]/);
    const departure = new Date(arr[0], arr[1] - 1, arr[2], arr[3], arr[4]);

    const diffMs = departure.valueOf() - today.valueOf();
    const days = getDays(diffMs);
    const hours = getHours(diffMs);
    const minutes = getMinutes(diffMs);
    if (diffMs <= 60000) return "now";
    return `in ${days}${hours}${minutes}`;
  };

  return (
    <Paragraph fontWeight={FontWeight.bold} fontSize={14} color={Colors.vBlue}>
      {getTimeUntilDeparture()}
    </Paragraph>
  );
};

export default TimeUntilDeparture;
