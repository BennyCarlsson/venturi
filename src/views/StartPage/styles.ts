import styled from 'styled-components';
import colors from 'tokens/colors';

export const StartPageWrapper = styled.div`
  background-color: ${colors.white};
`;
