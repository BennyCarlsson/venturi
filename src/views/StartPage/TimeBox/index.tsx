import Heading from "components/Heading";
import Paragraph from "components/Paragraph";
import Colors from "tokens/colors";
import { FontWeight } from "types";
import TimeUntilDeparture from "../TimeUntilDeparture";

import {
  TimeBoxWrapper,
  TimeWrapper,
  LocationWrapper,
  NumberBox,
  TimeLocationWrapper,
  NumberAndNameWrapper,
  StyledMapPinLogo,
  StyledRtTimeHeading,
} from "./styles";

type TimeBoxProps = {
  date?: string;
  time?: string;
  rtTime?: string;
  location?: string;
  number?: string;
  direction?: string;
};

const TimeBox = ({
  date,
  time,
  rtTime,
  location,
  number,
  direction,
  ...props
}: TimeBoxProps) => {
  const isOnTime = () => {
    if (!rtTime) return true;
    if (time !== rtTime) return false;
    return true;
  };

  const getTimeColor = () => {
    if (isOnTime()) return Colors.vBlue;
    return Colors.vRed;
  };

  return (
    <TimeBoxWrapper {...props}>
      <TimeLocationWrapper>
        <LocationWrapper>
          <StyledMapPinLogo />
          <Paragraph
            fontSize={14}
            fontWeight={FontWeight.bold}
            color={Colors.vBlue}
          >
            {location}
          </Paragraph>
        </LocationWrapper>
        <TimeWrapper>
          <Heading
            headingType={"h2"}
            fontWeight={FontWeight.bold}
            fontSize={48}
            color={getTimeColor()}
          >
            {rtTime ? rtTime : time}
          </Heading>
          {!isOnTime() && (
            <StyledRtTimeHeading
              headingType={"h3"}
              fontWeight={FontWeight.bold}
              fontSize={18}
            >
              {time}
            </StyledRtTimeHeading>
          )}
        </TimeWrapper>
        <TimeUntilDeparture date={date} time={rtTime ? rtTime : time} />
      </TimeLocationWrapper>
      <NumberAndNameWrapper>
        <NumberBox>
          <Paragraph
            fontWeight={FontWeight.normal}
            fontSize={14}
            color={Colors.white}
          >
            {number}
          </Paragraph>
        </NumberBox>
        <Paragraph fontWeight={FontWeight.medium} fontSize={14}>
          {direction}
        </Paragraph>
      </NumberAndNameWrapper>
    </TimeBoxWrapper>
  );
};

export default TimeBox;
