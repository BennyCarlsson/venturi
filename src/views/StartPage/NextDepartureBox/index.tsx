import Paragraph, { fontWeight } from 'components/Paragraph';
import colors from 'tokens/colors';
import { Departure } from 'types';
import { Box, Heading, Heading2, Heading3, Heading4 } from './styles';

const NextDepartureBox = (props: Departure) => {
  const { stop, time, rtTime, direction, sname } = props;
  return (
    <Box>
      <Heading>Departs</Heading>
      <Heading2>{stop}</Heading2>
      <Heading3>{time}</Heading3>
      <Heading4>{rtTime}</Heading4>
      <Paragraph
        color={colors.vGray}
        fontSize={14}
        fontWeight={fontWeight.normal}
      >{`${sname} ${direction}`}</Paragraph>
    </Box>
  );
};

export default NextDepartureBox;
