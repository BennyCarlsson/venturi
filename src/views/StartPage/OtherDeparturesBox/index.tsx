import Paragraph from 'components/Paragraph';
import { Legs } from 'types';
import { Box, DepartureWrapper } from './styles';

type Departures = {
  legs: Legs[];
};

const OtherDeparturesBox = (props: Departures) => {
  const { legs } = props;
  return (
    <Box>
      {legs.map((leg) => {
        return (
          <DepartureWrapper key={leg.Leg?.id}>
            <Paragraph>{leg.Leg?.Origin?.time}</Paragraph>
            <p>{leg.Leg.direction}</p>
            <p>{leg.Leg.sname}</p>
          </DepartureWrapper>
        );
      })}
    </Box>
  );
};

export default OtherDeparturesBox;
