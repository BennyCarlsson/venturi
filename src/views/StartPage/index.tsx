import { getTrip } from 'api/vasttrafik/vasttrafik';
import { Fragment, useCallback, useEffect, useState } from 'react';
import { ApiError, GetTripResponse, Legs } from 'types';
import DepartureBox from './DepartureBox';
import OtherDepartures from './OtherDeparturesBox';
import { StartPageWrapper } from './styles';

const spaldingsgatan = '9021014006160000';
const lillaBommen = '9021014004380000';
//const nordstan = '9021014004945000';
//const shouldShowRtTime = () => Origin?.rtTime !== Origin?.time;

const StartPage = () => {
  const [legs, setTrips] = useState<Legs[]>();
  const [error, setError] = useState<ApiError | null>(null);

  const getTimeTable = useCallback(async () => {
    const data = await getTrip(spaldingsgatan, lillaBommen);

    if ((data as ApiError).error) {
      console.log((data as ApiError).error);
    }

    const tripList = (data as GetTripResponse).TripList;
    if (tripList.error) {
      setError(data as ApiError);
    } else if (tripList) {
      const _legs = tripList.Trip as Legs[];
      console.log('tripList', tripList);
      console.log('_legs', _legs);
      setTrips(_legs);
    }
  }, []);

  useEffect(() => {
    getTimeTable();
  }, [getTimeTable]);

  return (
    <StartPageWrapper>
      {legs && legs.length > 0 && (
        <Fragment>
          <DepartureBox name={legs[0].Leg.Origin?.name} track={legs[0].Leg.Origin?.track} />
          <OtherDepartures legs={legs.slice(1, 4)} />
        </Fragment>
      )}
      {error && <h2>{error.error}</h2>}
      {error && <p>{error.errorText}</p>}
    </StartPageWrapper>
  );
};

export default StartPage;
