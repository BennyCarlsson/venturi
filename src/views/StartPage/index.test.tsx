import { render } from "@testing-library/react";
import StartPage from ".";
import { TripState } from "redux/tripSlice";
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { useSelector, useDispatch } from "react-redux";

jest.mock("react-redux", () => ({
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}));

const state: TripState = {
  loading: false,
  trips: [
    {
      id: "123",
      departureName: "name",
      departureTrack: "track",
      departureDate: "2021-05-08",
      departureTime: "15:30",
      departureNewTime: "15:31",
      destinationName: "destination name",
      number: "18",
      direction: "direction",
    },
    {
      id: "132",
      departureName: "name",
      departureTrack: "track",
      departureDate: "2021-05-08",
      departureTime: "15:35",
      departureNewTime: "15:38",
      destinationName: "destination name",
      number: "18",
      direction: "direction",
    },
    {
      id: "321",
      departureName: "name",
      departureTrack: "track",
      departureDate: "2021-05-08",
      departureTime: "15:30",
      departureNewTime: "15:31",
      destinationName: "destination name",
      number: "18",
      direction: "direction",
    },
    {
      id: "213",
      departureName: "name",
      departureTrack: "track",
      departureDate: "2021-05-08",
      departureTime: "15:30",
      departureNewTime: "15:31",
      destinationName: "destination name",
      number: "18",
      direction: "direction",
    },
  ],
};

describe("<StartPage />", () => {
  it("should match snapshot", () => {
    (useSelector as jest.Mock).mockReturnValueOnce(state.trips);
    (useDispatch as jest.Mock).mockReturnValue(() => {});
    const { asFragment } = render(<StartPage />);
    expect(asFragment()).toMatchSnapshot();
  });
});
