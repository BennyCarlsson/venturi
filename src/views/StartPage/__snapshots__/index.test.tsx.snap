// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StartPage /> should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-hHftDr bCBwfF"
  >
    <div
      class="sc-iBPRYJ HCnjh"
    >
      <div
        class="sc-fubCfw hIKVhW"
      >
        <span>
          <h1
            class="sc-bdfBwQ UZuHf"
            color="#FFFFFF"
            font-size="18"
            font-weight="500"
          >
            Departs from
          </h1>
        </span>
        <div
          class="sc-dQppl dWSjyx"
        >
          <svg
            class="sc-crrsfI gWLlXM"
          >
            train.svg
          </svg>
          <h2
            class="sc-gsTCUz htLLms"
            color="#FFFFFF"
            font-size="28"
            font-weight="500"
          >
            name
          </h2>
        </div>
      </div>
      <div
        class="sc-pFZIQ hxExeM"
      >
        <div
          class="sc-jrAGrp jzTLuk"
        >
          <p
            class="sc-gKsewC bMoIAp"
            color="#FFFFFF"
            font-size="14"
            font-weight="500"
          >
            Platform
          </p>
          <p
            class="sc-gKsewC iTDTZO sc-iqHYGH jyHRlY"
            color="#FFFFFF"
            font-size="24"
            font-weight="500"
          >
            track
          </p>
        </div>
        <div
          class="sc-kEjbxe bnNZBb"
        >
          <p
            class="sc-gKsewC bMoIAp"
            color="#FFFFFF"
            font-size="14"
            font-weight="500"
          >
            Destination name
          </p>
          <p
            class="sc-gKsewC iTDTZO sc-iqHYGH jyHRlY"
            color="#FFFFFF"
            font-size="24"
            font-weight="500"
          >
            Placeholder
          </p>
        </div>
      </div>
    </div>
    <div
      class="sc-dmlrTW dgOdrT"
    >
      <div
        class="sc-kfzAmx bUuPIZ"
      >
        <div
          class="sc-iwyYcG bbVDLA"
        >
          <svg
            class="sc-cxFLnm gbgjdP"
          >
            map-pin.svg
          </svg>
          <p
            class="sc-gKsewC inetcg"
            color="#719FB0"
            font-size="14"
            font-weight="700"
          >
            destination name
          </p>
        </div>
        <div
          class="sc-fKFyDc hkLWbf"
        >
          <h2
            class="sc-gsTCUz fNVbJl"
            color="#A0616A"
            font-size="48"
            font-weight="700"
          >
            15:31
          </h2>
          <h3
            class="sc-dlfnbm gDZnXS sc-bBXqnf jEPXpx"
            font-size="18"
            font-weight="700"
          >
             15:30
          </h3>
        </div>
        <p
          class="sc-gKsewC inetcg"
          color="#719FB0"
          font-size="14"
          font-weight="700"
        >
          now
        </p>
      </div>
      <div
        class="sc-lmoMRL bYrWNd"
      >
        <div
          class="sc-iJuUWI fytIuD"
        >
          <p
            class="sc-gKsewC gtcvBk"
            color="#FFFFFF"
            font-size="14"
            font-weight="400"
          >
            18
          </p>
        </div>
        <p
          class="sc-gKsewC eCnwPF"
          font-size="14"
          font-weight="500"
        >
          direction
        </p>
      </div>
    </div>
    <div
      class="sc-bqyKva jSOVnk"
    >
      <a
        class="sc-kstrdz cQnKCZ"
      >
        <span />
        <p
          class="sc-gKsewC ecZMIM sc-hBEYos"
          color="#FFFFFF"
          font-size="18"
          font-weight="500"
        >
          Other departures
        </p>
        <svg
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polyline
            points="6 9 12 15 18 9"
          />
        </svg>
      </a>
      <div
        class="sc-fodVxV faFDam"
        style="height: 0px;"
      >
        <div
          class="sc-fFubgz gqHger"
        >
          <div
            class="sc-bkzZxe gAHmgU"
          >
            <p
              class="sc-gKsewC cQZNpE"
              color="#FFFFFF"
              font-size="18"
            >
              15:35
            </p>
            <p
              class="sc-gKsewC cQZNpE sc-idOhPF cbscxb"
              color="#FFFFFF"
              font-size="18"
            >
              direction
            </p>
          </div>
          <div
            class="sc-dIUggk bpjTIx"
          >
            <p
              class="sc-gKsewC dfHdoQ"
              font-size="14"
            >
              18
            </p>
          </div>
        </div>
        <div
          class="sc-fFubgz gqHger"
        >
          <div
            class="sc-bkzZxe gAHmgU"
          >
            <p
              class="sc-gKsewC cQZNpE"
              color="#FFFFFF"
              font-size="18"
            >
              15:30
            </p>
            <p
              class="sc-gKsewC cQZNpE sc-idOhPF cbscxb"
              color="#FFFFFF"
              font-size="18"
            >
              direction
            </p>
          </div>
          <div
            class="sc-dIUggk bpjTIx"
          >
            <p
              class="sc-gKsewC dfHdoQ"
              font-size="14"
            >
              18
            </p>
          </div>
        </div>
        <div
          class="sc-fFubgz gqHger"
        >
          <div
            class="sc-bkzZxe gAHmgU"
          >
            <p
              class="sc-gKsewC cQZNpE"
              color="#FFFFFF"
              font-size="18"
            >
              15:30
            </p>
            <p
              class="sc-gKsewC cQZNpE sc-idOhPF cbscxb"
              color="#FFFFFF"
              font-size="18"
            >
              direction
            </p>
          </div>
          <div
            class="sc-dIUggk bpjTIx"
          >
            <p
              class="sc-gKsewC dfHdoQ"
              font-size="14"
            >
              18
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
